### PR TITLE
refactor StyledBuffer

### PR DESCRIPTION
**Pull request by klensy**
Refactors StyledBuffer `text` and `styles` fields content into StyledChar and touches some other stuff.

### CODE REVIEW

**Comment by klensy** [comment]
I expected that change in `prepend` will alter output, as in some cases it will not generate additional spaces, but actually nothing changed! Looks like somewhere deep in emitter additional spaces was cleaned before.

File: compiler/rustc_errors/src/styled_buffer.rs [hunk 1]
```diff
@@ -1,39 +1,56 @@
...
+
+impl StyledChar {
+    fn new(chr: char, style: Style) -> Self {
+        StyledChar { chr, style }
+    }
+}
+
+impl Default for StyledChar {
+    fn default() -> Self {
+        StyledChar::new(' ', Style::NoStyle)
```
Comment (lcnr) [inline compiler/rustc_errors/src/styled_buffer.rs:24]: this differs from the default for `char`, is this intentional?

File: compiler/rustc_errors/src/styled_buffer.rs [hunk 2]
```diff
@@ -49,29 +66,31 @@ impl StyledBuffer {
     }
 
     fn ensure_lines(&mut self, line: usize) {
-        while line >= self.text.len() {
-            self.text.push(vec![]);
-            self.styles.push(vec![]);
+        while line >= self.lines.len() {
+            self.lines.push(vec![]);
```
Comment (lcnr) [inline compiler/rustc_errors/src/styled_buffer.rs:70]:
```suggestion
        if self.lines.len() <= line {
            self.lines.resize(line + 1);
```

File: compiler/rustc_errors/src/styled_buffer.rs [hunk 3]
```diff
@@ -49,29 +66,31 @@ impl StyledBuffer {
...
         }
     }
 
+    /// Sets `chr` with `style` for given `line`, `col`.
+    /// If line not exist in `StyledBuffer`, adds lines up to given
+    /// and fills last line with spaces and `Style::NoStyle` style
```
Comment (lcnr) [inline compiler/rustc_errors/src/styled_buffer.rs:76]:
```suggestion
    /// If `line` does not exist in our buffer, adds empty lines up to the given
    /// and fills the last line with unstyled whitespace.
```
maybe something like this reads slightly better :thinking:

File: compiler/rustc_errors/src/styled_buffer.rs [hunk 4]
```diff
@@ -49,29 +66,31 @@ impl StyledBuffer {
...
     pub fn putc(&mut self, line: usize, col: usize, chr: char, style: Style) {
         self.ensure_lines(line);
-        if col < self.text[line].len() {
-            self.text[line][col] = chr;
-            self.styles[line][col] = style;
+        if col < self.lines[line].len() {
```
Comment (lcnr) [inline compiler/rustc_errors/src/styled_buffer.rs:79]: can we replace that large if statement with
```suggestion
        if col >= self.lines[line].len() {
            self.lines[line].resize(col + 1, StyledChar::new(' ', Style::NoStyle);
        }
        self.lines[line][col] = StyledChar::new(chr, style);
```

File: compiler/rustc_errors/src/styled_buffer.rs (hunk 1 above)
Comment (lcnr) [reply compiler/rustc_errors/src/styled_buffer.rs:24]: i think I would prefer to add an associated constant instead, so something like
```rust
impl StyledChar {
    const SPACE: Self = StyledChar::new(' ', Style::NoStyle);
}
```
that requires you to make `StyledChar::new` a const fn

**Review by lcnr — APPROVED** [review]
some small nits, feel free to implement the ones you consider an improvement, afterwards r=me

File: compiler/rustc_errors/src/styled_buffer.rs (hunk 1 above)
Comment (klensy) [reply compiler/rustc_errors/src/styled_buffer.rs]: Fixed.

File: compiler/rustc_errors/src/styled_buffer.rs (hunk 2 above)
Comment (klensy) [reply compiler/rustc_errors/src/styled_buffer.rs]: Fixed with resize_with.

File: compiler/rustc_errors/src/styled_buffer.rs (hunk 3 above)
Comment (klensy) [reply compiler/rustc_errors/src/styled_buffer.rs]: Fixed.

File: compiler/rustc_errors/src/styled_buffer.rs (hunk 4 above)
Comment (klensy) [reply compiler/rustc_errors/src/styled_buffer.rs]: Fixed.

File: compiler/rustc_errors/src/styled_buffer.rs [hunk 5]
```diff
@@ -49,29 +62,25 @@ impl StyledBuffer {
     }
 
     fn ensure_lines(&mut self, line: usize) {
-        while line >= self.text.len() {
-            self.text.push(vec![]);
-            self.styles.push(vec![]);
+        if line >= self.lines.len() {
+            self.lines.resize_with(line + 1, || vec![]);
```
Comment (lcnr) [inline compiler/rustc_errors/src/styled_buffer.rs:66]:
```suggestion
            self.lines.resize(line + 1, Vec::new());
```
this is a really small thing, but i quite strongly dislike `resize_with(line + 1, || vec![])`.

`Vec::new()` doesn't allocate and should be fully evaluated at compile time. So there is no perf benefit to using `resize_with` and `resize` is easier to read

**Review by lcnr — CHANGES_REQUESTED** [review]
a final small stylistic nit, sorry '^^

@bors delegate+

**Comment by lcnr** [comment]
@bors r+ rollup

forgot that `bors` commands don't work in review comments :sweat_smile:

**Pull request merged**